### ISSUE TITLE
Blazor doesn't render tags in single line markup(@:) correctly

### ISSUE THREAD

**Issue by ajaybhargavb** [aspnet/AspNetCore#11612]
**Repro:**

Input:
```cshtml
@if (test)
{
    @:Hello <span>@text</span>
}

@code {
    string text = "world!";
    bool test = true;
}
```
Generated code:
```cs
builder.AddMarkupContent(1, "    Hello <span>");
builder.AddContent(2, text);
builder.AddMarkupContent(3, "</span>\r\n");
```

Expected Output:
```html
Hello <span>world!</span>
```

Actual Output:
```html
Hello <span></span> world!
```

**Comment by ajaybhargavb** [comment]
@rynowak @SteveSandersonMS, this doesn't look like a Razor compiler issue because the generated code makes sense to me. Do we do some auto tag completion magic in Blazor runtime that messes up how it is rendered?

**Comment by SteveSandersonMS** [comment]
When we ask the browser to insert markup, we first tell it to parse that markup. Browsers do funky stuff to "fix" your mistakes in HTML, so for example `    Hello <span>` probably gets parsed as if you wrote `    Hello <span></span>`. Likewise it probably ignores the `</span>` because that's not valid on its own.

To avoid this weirdness, we are meant to be requiring that the markup is well-formed and the tags are correctly balanced. `</span>` on its own should *not* be allowed by the compiler.

It sounds like the fix is to disallow `@:` syntax, right? There's no use for it in `.razor`; it's just harmful.

**Comment by ajaybhargavb** [comment]
> It sounds like the fix is to disallow @: syntax, right? There's no use for it in .razor; it's just harmful.

I'm not sure if there is no use for `@:` in `.razor`. Without it you won't be able to transition out of C# in code block without wrapping it in a tag. This was also an issue with `<text>` tags which is special in Razor code blocks. This was originally reported here https://github.com/aspnet/AspNetCore/issues/11316.
@NTaylorMullen, can you think of other ways to do what `@:` currently lets you do?

**Comment by NTaylorMullen** [comment]
Using `<text>` on its own should suffice; that being said you are definitely restricted from utilizing a tag named `<text>` directly in a code block without `@:`. To work around that users can store their HTML in variables and spit them out if they need the text tag. I know it's not ideal but there's not a great alternative.

The other option would be to make `@:` parse HTML properly which would be difficult but more importantly would differ from the normal Razor syntax where people 99% of the time use it to render unbalanced HTML in different code blocks (not supported in component files):
```
@if (something)
{
    @: <p>
}

...

@if (something)
{
    @:</p>
}
```

So i'm on board with disallowing `@:`  in Component files, at least for now with a specific error. Can always add a resemblance of the feature functionality later.

**Comment by git-net** [comment]
It is theoretically correct。but when we port our old code，the same razor  code ，different。

**Comment by mrlife** [comment]
Are we sure this is a duplicate of https://github.com/dotnet/aspnetcore/issues/20971?

**Comment by MithrilMan** [comment]
I agree that #20971 should be reopened, in my scenario I encountered the bug in this code 

```
private RenderFragment RenderRows(IEnumerable<(string label, string value)> rows) =>
@:@{
   foreach (var row in rows) {
      <span class="font-bold">@Translate(row.label)</span>
      <span class="">@row.value</span>
   }}
;
```

last `;` had to be put on its own lane or the compiler complains.... (.Net 5)
Does .Net6 has improvement in this scenarios?

**Comment by NTaylorMullen** [comment]
@MithrilMan sorry for the confusion, would you mind filing another issue in dotnet/aspnetcore? That's where the Razor compiler exists these days. We did a huge repo separation just a little while back 😄

**Comment by MithrilMan** [comment]
mmm actually this issue is already on dotnet/aspnetcore

**Comment by NTaylorMullen** [comment]
> mmm actually this issue is already on dotnet/aspnetcore

woops, well there we go that's my confusion 😄
